### ISSUE TITLE
US Tax rules are year specific

### ISSUE THREAD

**Issue by daveloyall** [b-k/py1040#1]
Please change any reference to `1040` to `1040 (2015)`.  This will help the code grow next year.

**Comment by b-k** [comment]
Yes, the documentation should be clear on this. Thanks for fixing readme.md.

As of version 42d0cdd , this project depends on the form database in the 1040.js project. The intent is to develop a reference form set for use regardless of front-end or even language choice.

My hope for that reference set was that we'd use git tags or branches to indicate the year, so the forms themselves wouldn't have any indication of what year they're for. Am entirely open to suggestions on whether this setup makes sense, or how it can be better.

**Comment by ytjohn** [comment]
I think it would be better to maintain a 1040 for each year. If someone wanted to go back and recalculate a prior year's taxes, a command line option or input could select the right year. 

Maybe you could develop a base set of rules and put year-specific deviations in a script, but I think it would be easier to just modify a new 1040 each year.

**Comment by daveloyall** [comment]
@ytjohn is on the right track.  The IRS itself considers each year's 1040 (and other forms) to be a distinct entity from prior years' copies of the same form.  See [search results for "1040" on the IRS Forms and Publications site](https://apps.irs.gov/app/picklist/list/priorFormPublication.html?value=1040&criteria=formNumber&submitSearch=Find) for example.

**Comment by b-k** [comment]
OK, I've tried a few approaches to putting year markers on tax forms without using revision control, and I'm not happy with them.

Tl;dr: putting multiple years' forms in the same place gets into weeds I'm disinclined to get into on this project, and using revision control to jump versions or compare versions is a natural application of revision control.

The first option is to add `IF(year=>2016, use_this_new_cell, else_use_this_older_cell)` constructions for those cells that changed across years. This is likely not too cluttered, and given the rate of change in the tax forms is reasonably easy to follow.

But we'd need to worry about how dependency graphs get generated: is 1040 line 45 (AMT) dependent on line 35 of f6251-2015, f6251-2016, f6251-2017, ..., or do we find a way to ignore all off-years when builiding the dependency graph? This is a not-insurmountable problem, but isn't trivial either, given that the dependency calculator right now is just a macro that looks for certain text patterns and then generates code to evaluate all dependent cells before evaluating the current one.

Another alternative is to replicate 100% of the cells for every year, even though 98% will be identical year-to-year. Now we have parallel years and don't have to think hard about the dependency tree branching. This simplification comes at the cost of a ton of redundancy, which creates the usual redundancy problems (and is just æsthetically unappealing).

At that point, it makes as much sense to just put those alternatives in separate branches.  Given the 2015 forms in a 2015 branch (which they now are in 1040.js), you can see the difference in the forms via `git diff master 2015`.  To run your taxes for 2015, check last year's branch and rebuild the forms, which is basically two lines on the command line (`cd 1040.js; git checkout 2015; cd ..; make taxforms.py`), either typed in or called at the beginning of the Python script given a switch.
